Add Bottom to LabelPosition for Cartesian Charts

L-Andrade opened the pull request:
Adds `Bottom` as a `LabelPosition`, which positions at the bottom of the graph. Basically the inverse of `Top`.
Solves https://github.com/patrykandpatrick/vico/discussions/701

| Bottom | Top |
| --- | --- |
| <img width="400" alt="Screenshot 2024-05-22 at 14 09 20" src="https://github.com/patrykandpatrick/vico/assets/20290474/26e98c4e-a234-49e7-99df-286a7ceb3989"> | <img width="400" alt="Screenshot 2024-05-22 at 14 24 56" src="https://github.com/patrykandpatrick/vico/assets/20290474/d742a335-3d78-4cf1-97c2-6bebaabbd111"> |

patrickmichalik reviewed (changes requested):
Hello, and thanks for the PR! When `LabelPosition.Bottom` is applied, `DefaultCartesianMarker` should request a bottom inset to ensure that there’s enough room for the label. There may not be a bottom axis, or the height of the label may exceed that of the bottom axis. Could you update `getInsets` (line 189) accordingly? This modification won’t change the end result in instances where there’s already sufficient room.

L-Andrade commented:
Hey @patrickmichalik, thanks for the feedback 😄 

> Hello, and thanks for the PR! When `LabelPosition.Bottom` is applied, `DefaultCartesianMarker` should request a bottom inset to ensure that there’s enough room for the label. There may not be a bottom axis, or the height of the label may exceed that of the bottom axis. Could you update `getInsets` (line 189) accordingly? This modification won’t change the end result in instances where there’s already sufficient room.

Ah, missed that! Good point, wouldn't look good without that. I've tested it and it works nicely after this commit: f56e3fdfa8f0a1249e94b38c395944566cb85b8d

Here's a couple screenshots, I've added some background colors so we can easily identify the chart's size
| Bottom | Top |
| --- | --- |
| <img width="400" alt="Screenshot 2024-05-23 at 16 38 03" src="https://github.com/patrykandpatrick/vico/assets/20290474/7f4db3f0-ab21-4c59-9d01-d7277a4733b7"> | <img width="400" alt="Screenshot 2024-05-23 at 16 38 35" src="https://github.com/patrykandpatrick/vico/assets/20290474/ddd998cc-3998-4408-8ad3-04d41067669c"> |

It took me a while to notice that `Marker.kt`'s marker had an override for `getInsets`. Since the code was basically the same as the `getInsets` except for the addition of the shadow clipping, I changed it to call the `super`'s `getInsets` and to add the clipping after, which makes it easier to test other `LabelPosition`s' behaviour.

Changing the code to use `when` should also make it easier in the future if there's a need for other label positions, as they will require it to be exhaustive (for example, if there's a need to add `BelowPoint` in the future - it won't compile until we update it in all these places now)

Let me know if you'd prefer the previous implementation there (in `Marker.kt`), since it is just sample code. I'll be happy to roll it back if needed.

patrickmichalik commented on sample/src/main/java/com/patrykandpatrick/vico/sample/showcase/Marker.kt:
```diff
@@ -101,14 +101,18 @@ internal fun rememberMarker(
                 with(context) {
-                    outInsets.top =
+                    super.getInsets(context, outInsets, horizontalDimensions)
+                    val shadowClipping =
```
I’d suggest using a more descriptive name here. Clipping is what we’re looking to avoid. Something along the lines of `shadowPadding` or `shadowInset` would, I think, better reflect the variable’s purpose.

patrickmichalik reviewed (changes requested):
Thanks, @L-Andrade! The changes that you mention look good. I have three minor suggestions.

L-Andrade commented on sample/src/main/java/com/patrykandpatrick/vico/sample/showcase/Marker.kt:
```diff
@@ -101,14 +101,18 @@ internal fun rememberMarker(
+                        LabelPosition.Top -> outInsets.top += shadowClipping
+                        LabelPosition.Bottom -> outInsets.bottom += shadowClipping
+                        LabelPosition.AroundPoint, LabelPosition.AbovePoint -> Unit
+                    }
```
Oops, missed that. I've fixed it here: f7a95bc5581be840c144fae639ab23cb26a33c2c

patrickmichalik replied:
Thanks!

patrickmichalik commented on sample/src/main/java/com/patrykandpatrick/vico/sample/showcase/Marker.kt:
```diff
@@ -101,14 +101,18 @@ internal fun rememberMarker(
                             CLIPPING_FREE_SHADOW_RADIUS_MULTIPLIER * LABEL_BACKGROUND_SHADOW_RADIUS_DP -
                                 LABEL_BACKGROUND_SHADOW_DY_DP
                         )
                             .pixels
```
Thanks!

patrickmichalik commented on vico/core/src/main/java/com/patrykandpatrick/vico/core/cartesian/marker/DefaultCartesianMarker.kt:
```diff
@@ -181,8 +191,19 @@ public open class DefaultCartesianMarker(
+                }
+
+            LabelPosition.AroundPoint -> Unit // Will be inside the chart
+        }
```
Thanks!

patrickmichalik reviewed (approved):
Cheers! @Gowsky is on a business trip at the moment, but he told me before leaving that he didn’t have any further suggestions, so we’re ready to go.

patrickmichalik commented:
For completeness, I wanted to note that I missed a minor issue [here](https://github.com/patrykandpatrick/vico/pull/721#discussion_r1614776991). Whatever the condition for the bottom shadow inset being added may be, its size should be different than that of the top shadow inset. This is because the shadow is always shifted downward by 2 dp. I erroneously concluded that two different sizes would be required only once the out-of-scope change mentioned is made—perhaps because I flipped the shadow along with the label background in my head. I hadn’t yet merged the PR when I realized this, but it’s a small thing, and it’s my mistake for overlooking it in the review. I’ll take care of it when updating the logic to always add both shadow insets.

L-Andrade commented:
@patrickmichalik Sorry for missing it too - makes sense. Let me know if you need anything from my side!

patrickmichalik commented:
No worries, @L-Andrade! Such details sometimes slip through the cracks. I’ve fixed the issue [here](https://github.com/patrykandpatrick/vico/commit/0e2f778cc8483eddbb5726d22ebd24bb4179aabd#diff-62ed644d764931f84462a61244a8ec1a0d77a99dde2a17e2d1b6b96f9fae143b). Thanks again!
